RZAssemblage/RZAssemblage.m

```diff
@@ -0,0 +1,289 @@
+        }
+    }
+    self.updateCount += 1;
+}
+
+- (void)closeBatchUpdate
+{
+    self.updateCount -= 1;
+    if ( self.updateCount == 0 ) {
+        RZAssemblageLog(@"Change:%@ -> %p:\n%@", self, self.delegate, self.changeSet);
+        [self.delegate assemblage:self didEndUpdatesWithChangeSet:self.changeSet];
+        self.changeSet = nil;
+    }
+}
+
+#pragma mark - RZAssemblageDelegate
+
+- (void)willBeginUpdatesForAssemblage:(RZAssemblage *)assemblage
+{
+    [self openBatchUpdate];
+}
+
+- (void)notifyObjectUpdate:(id)anObject
+{
+    NSParameterAssert(anObject);
+    NSUInteger index = [self elementsIndexOfObject:anObject];
+    RZAssemblageLog(@"%p:Update %@ at %zd", self, anObject, index);
+    NSAssert(index != NSNotFound, @"Object is not part of assemblage");
+    [self openBatchUpdate];
+    [self.changeSet updateAtIndexPath:[NSIndexPath indexPathWithIndex:index]];
+    [self closeBatchUpdate];
+}
+
+- (void)assemblage:(RZAssemblage *)assemblage didEndUpdatesWithChangeSet:(RZAssemblageChangeSet *)changeSet
+{
+    RZRaize(self.changeSet != nil, @"Must begin an update on the parent assemblage before mutating a child assemblage");
+    NSUInteger assemblageIndex = [self indexOfAssemblage:assemblage];
+    [self.changeSet mergeChangeSet:changeSet withIndexPathTransform:^NSIndexPath *(NSIndexPath *indexPath) {
+        return [indexPath rz_indexPathByPrependingIndex:assemblageIndex];
+    }];
+    [self closeBatchUpdate];
+}
+
+#pragma mark - Index Path Helpers
+
+- (void)insertObject:(id)object atIndexPath:(NSIndexPath *)indexPath
+{
+    NSMutableArray *proxy = [[self assemblageAtIndexPath:[indexPath indexPathByRemovingLastIndex]] mutableChildren];
+    [proxy insertObject:object atIndex:[indexPath rz_lastIndex]];
+}
+
+- (void)removeObjectAtIndexPath:(NSIndexPath *)indexPath
+{
+    NSMutableArray *proxy = [[self assemblageAtIndexPath:[indexPath indexPathByRemovingLastIndex]] mutableChildren];
+    [proxy removeObjectAtIndex:[indexPath rz_lastIndex]];
+}
+
+- (void)moveObjectAtIndexPath:(NSIndexPath *)fromIndexPath toIndexPath:(NSIndexPath *)toIndexPath
+{
+    NSMutableArray *fproxy = [[self assemblageAtIndexPath:[fromIndexPath indexPathByRemovingLastIndex]] mutableChildren];
+    NSMutableArray *tproxy = [[self assemblageAtIndexPath:[toIndexPath indexPathByRemovingLastIndex]] mutableChildren];
+
+    NSObject *object = [fproxy objectAtIndex:[fromIndexPath rz_lastIndex]];
+    [fproxy removeObjectAtIndex:[fromIndexPath rz_lastIndex]];
+    [tproxy insertObject:object atIndex:[toIndexPath rz_lastIndex]];
+}
+
+@end
+
+@implementation RZAssemblage (Protected)
+
+- (NSUInteger)countOfElements
+{
+    RZSubclassMustImplement(NSNotFound);
+}
+
+- (id)objectInElementsAtIndex:(NSUInteger)index
+{
+    RZSubclassMustImplement(nil);
+}
+
+- (RZAssemblage *)nodeAtIndex:(NSUInteger)index
+{
+    RZSubclassMustImplement(nil);
+}
+
+- (NSUInteger)indexOfAssemblage:(RZAssemblage *)assemblage
```

Comment:
Just thinking `indexPathOfObjectPassingTest:` or `indexOfObjectPassingTest:` would be nice to implement on assemblage.